src/bin/nix-channel-index.rs

```diff
@@ -0,0 +1,298 @@
+
+        let all_paths = systems.into_iter().flat_map(|system| {
+            // These are the paths that show up in `nix-env -qa`.
+            let normal_paths =
+                nixpkgs::query_packages(&args.nixpkgs, None, args.show_trace, system);
+
+            // We also add some additional sets that only show up in `nix-env -qa -A someSet`.
+            //
+            // Some of these sets are not build directly by hydra. We still include them here
+            // since parts of these sets may be build as dependencies of other packages
+            // that are build by hydra. This way, our attribute path information is more
+            // accurate.
+            //
+            // We only need sets that are not marked "recurseIntoAttrs" here, since if they are,
+            // they are already part of normal_paths.
+
+            normal_paths.chain(EXTRA_SCOPES.iter().flat_map(move |scope| {
+                nixpkgs::query_packages(&args.nixpkgs, Some(scope), args.show_trace, system)
+            }))
+        });
+
+        let paths: Vec<StorePath> = all_paths
+            .map(|x| x.chain_err(|| ErrorKind::QueryPackages))
+            .collect::<Result<_>>()?;
+
+        fetch_file_listings(&fetcher, args.jobs, paths)
+    };
+
+    // Treat request errors as if the file list were missing
+    let files = files.map(|r| {
+        r.unwrap_or_else(|e| {
+            errst!("\n{}", e.display_chain());
+            None
+        })
+    });
+
+    // Add progress output
+    let (mut indexed, mut missing) = (0, 0);
+    let files = files.inspect(|entry| {
+        if entry.is_some() {
+            indexed += 1;
+        } else {
+            missing += 1;
+        };
+
+        errst!("+ generating index: {:05} paths found :: {:05} paths not in binary cache :: {:05} paths in queue \r",
+               indexed, missing, watch.queue_len());
+        io::stderr().flush().expect("flushing stderr failed");
+    });
+
+    let mut files = files.filter_map(future::ready);
+
+    errst!("+ generating index");
+    errst!("\r");
+
+    while let Some((path, nar, files)) = files.next().await {
+        let origin = path.origin();
+
+        for item in files.to_list(&[]) {
+            if let FileNode::Regular {
+                size: _,
+                executable: true,
+            } = item.node
+            {
+                let path = PathBuf::from(OsString::from_vec(item.path));
+
+                if let Ok(binary) = path.strip_prefix("/bin") {
+                    let attr = origin.attr.clone();
+                    let system = origin.system.clone();
+                    let binary: String = binary.to_string_lossy().into();
+
+                    if !binary.starts_with('.') {
+                        connection
+                            .execute(
+                                "insert or replace into Programs(name, system, package) values (?, ?, ?)",
+                                (attr, system, binary),
```

Comment:
Is this correct? I don't know how the current programs sqlite db works but from the naming I would have expected "package" to be the attr and "name" the binary name?

Reply:
Good find. Will fix in a bit.